src/operator/softmax_output-inl.h

```diff
@@ -118,6 +153,28 @@ class SoftmaxOutputOp : public Operator {
         SoftmaxGrad(grad, out, label);
       }
       grad *= DType(param_.grad_scale);
+
+      // normalize gradient
+      if (param_.normalization == softmaxout_enum::kBatch) {
+        grad /= static_cast<float>(grad.size(0));
+      } else if (param_.normalization == softmaxout_enum::kValid) {
+        int ignore = static_cast<int>(param_.ignore_label);
+        int64_t valid_cnt = 0;
+        float scale = 1.0f;
+        Tensor<cpu, 1, DType> workspace =
+          ctx.requested[softmaxout_enum::kTempSpace].get_space_typed<cpu, 1, DType>(label.shape_,
+                                                                                    NULL);
+        Copy(workspace, label, label.stream_);
+        for (index_t i = 0; i < workspace.size(0); ++i) {
```

Comment:
@piiswrong it is used in training with sequences with different lengths. This option is used to drop padded frames.